keps/sig-node/3008-cri-class-based-resources/README.md

```diff
@@ -0,0 +1,1598 @@
+would implement restrictions based on the namespace.
+
+```diff
+ // ResourceQuotaSpec defines the desired hard limits to enforce for Quota.
+ type ResourceQuotaSpec struct {
+     // hard is the set of desired hard limits for each named resource.
+     Hard ResourceList
+     // A collection of filters that must match each object tracked by a quota.
+     // If not specified, the quota matches all objects.
+     Scopes []ResourceQuotaScope 
+     // scopeSelector is also a collection of filters like scopes that must match each
+     // object tracked by a quota but expressed using ScopeSelectorOperator in combination
+     // with possible values.
+     ScopeSelector *ScopeSelector
++    // PodClassResources contains the allowed pod-level class resources.
++    PodClassResources []ClassResourceInfo
++    // ContainerClassResources contains the allowed container-level class resources.
++    ContainerClassResources []ClassResourceInfo
+ }
+
+ // ResourceQuotaStatus defines the enforced hard limits and observed use.
+ type ResourceQuotaStatus struct {
+        ...
+        // Used is the current observed total usage of the resource in the namespace
+        // +optional
+        Used ResourceList
++       // PodClassResources contains the enforced set of pod-level class resources available.
++       PodClassResources []ClassResourceInfo
++       // ContainerClassResources contains the enforced set of container class resources available.
++       ContainerClassResources []ClassResourceInfo
+ }
+
+
+```
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+We extend the CRI protocol to contain information about the QoS-class
+resource assignment of containers and pods. Resource assignment requests will
+be simple key-value pairs (*resource-type=class-name*)
+
+Container runtime is expected to be aware of all resource types and the classes
+within. The CRI protocol is extended to be able to communicate the available
+QoS-class resources from the runtime to the client. This information includes:
+- Available QoS-class resource types.
+- Available classes within each resource type.
+- Whether the resource type is immutable or if it supports in-place updates.
+  In-place updates of resoures might not be possible because of runtime
+  limitations or the underlying technology, for example.
+
+Pod-level and container-level QoS-class resources are completely independent
+resource types. E.g. specifying something in the pod-level request does not
+mean specifying a pod-level default for all containers of the pod.
+
+Currently we identify two types of container-level QoS-class resources (RDT and
```

Comment:
Is there any desire to add tests or anything to standardize behavior here? Since Kubelet is not aware, I'm not sure if there should be tests in kubernetes repo, but then again, we make the claim blockio and rdt are supported by both runtimes. If there's a future runtime, would a user expect those to work? In other words, are these class resources an implicit requirement of CRI conformance? I'm fine if we put that in open questions

Reply:
agree and is probably a next phase item..

Reply:
> Is there any desire to add tests or anything to standardize behavior here? Since Kubelet is not aware, I'm not sure if there should be tests in kubernetes repo, but then again, we make the claim blockio and rdt are supported by both runtimes. If there's a future runtime, would a user expect those to work? In other words, are these class resources an implicit requirement of CRI conformance? I'm fine if we put that in open questions

Good question, indeed. This is something I've been struggling a bit. I.e. what should/could we state about RDT and blockio in the KEP. In principle this (QoS-class resources) should be generic mechanism, I think. (compare e.g. to extended resources / device plugins, nothing is stated about what kind of resources should be available). OTOH, now this proposal states that we would add some consts (`rdt` and `blockio`) to k8s.io/cri-api 🙃

Also, the API changes are platform agnostic (as @sftim just commented in #3008) but the currently available "qos-class resource implementations" in runtimes are platform-specific. Especially blockio seems very unlikely to see a Windows implementation.

Thus, I'd be inclined to keep rdt and blockio out of the specification in any way. Just use them as practical examples. WDYT?

Reply:
This should be now resolved by the "official" well-known QoS resource names that will be specified in the API at later phase

Reply:
@haircommander marking this as resolved